### PR TITLE
modified TableDescription formatting

### LINKED ISSUE #459: sqlcmd profile improvements
So the table profiling is a bit unclear on how to read the output and what to do with nan values.
For instance, running on the table below, it could be `nan` for many reasons, wrong datatypes, db doesn't support the unique functions (STDEV function), confusing values (`2.456e+01`) etc.

We should improve the guide and the functionality of it, and produce a result that's similar in essence to the pandas profiling report.
Screen Shot 2023-04-26 at 3 57 46 PM

@edublancas feel free to add more context.

### PR DESCRIPTION
**Pull request by AnirudhVIyer**
## Describe your changes
Modified the `TableDescription` Class such that now we
1) check the data for each column ( numeric or categorical)
2) check a column contains numerical data in str form (this creates Nan issues) and issue a warning
3) Calculate Mean, Min, Max, Count, STD, percentiles for numeric data
4) Calculate Unique, Top, Frequency for categorical data
5) Apply styling to indicate columns with data mismatch 
6) Added formatting so we don't encounter scientific values `eg. 1.621e+1`

Now the `profile` command acts similar to the pandas `describe` function
### Jupysql profile
Screenshot 2023-06-13 at 4 34 15 PM

### Pandas profile
Screenshot 2023-06-13 at 4 34 28 PM## Issue number

Closes #459 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--616.org.readthedocs.build/en/616/

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204482537184202

### CODE REVIEW

**Review by edublancas — CHANGES_REQUESTED** [review]
looks great overall!

can we change the red background for yellow? Also, let's add some message below the table explaining why we have the yellow background.

something like "columns A, B, C are categorical, hence median, mean, cannot be computed"

**Comment by AnirudhVIyer** [comment]
> looks great overall!
> 
> can we change the red background for yellow? Also, let's add some message below the table explaining why we have the yellow background.
> 
> something like "columns A, B, C are categorical, hence median, mean, cannot be computed"

Sure! making the changes

**Comment by AnirudhVIyer** [comment]
@edublancas made the changes 

Screenshot 2023-06-14 at 2 14 06 PM

**Comment by edublancas** [comment]
please merge from master, I think the error you're getting got fixed already

@tonykploomber: is the error caused by the mariadb image problem you encountered? because the logs suggest the problem is with oracle

**Comment by jinniw43805** [comment]
Seems the [latest build](https://github.com/ploomber/jupysql/actions/runs/5281544493/jobs/9555286307) with this oracledb error too, however it's not impacted to the testing flow.

I believe rebasing with latest master will solve the issue, however, something we need to check about oracledb

**Comment by tonykploomber** [comment]
Seems the [latest build](https://github.com/ploomber/jupysql/actions/runs/5281544493/jobs/9555286307) with this oracledb error too, however it's not impacted to the testing flow.

I believe rebasing with latest master will solve the issue, however, something we need to check about oracledb

**Comment by AnirudhVIyer** [comment]
I will rebase and try again.

File: src/sql/inspect.py [hunk 1]
```diff
@@ -77,6 +77,51 @@ def _get_row_with_most_keys(rows):
...
+def _is_numeric(value):
+    """Check if a column has numeric and not categorical datatype"""
+    try:
+        float(value)  # Try to convert the value to float
+        return True
+    except ValueError:
+        return False
+
+
+def _check_wrong_datatype(column, value):
```
Comment (neelasha23) [inline src/sql/inspect.py:89]: rename -> `_is_numeric_as_str`

File: src/sql/inspect.py [hunk 2]
```diff
@@ -77,6 +77,51 @@ def _get_row_with_most_keys(rows):
     return list(rows[max_idx])
 
 
+def _is_numeric(value):
+    """Check if a column has numeric and not categorical datatype"""
```
Comment (neelasha23) [inline src/sql/inspect.py:81]: We should add test cases for these functions

File: src/sql/inspect.py [hunk 3]
```diff
@@ -77,6 +77,51 @@ def _get_row_with_most_keys(rows):
...
+        """
+    return f"<style>{styles}</style>"
+
+
+def _generate_message(column_indices, columns):
+    """Generate a message indicating all columns with a datatype mismatch"""
+    message = "Columns "
+    for c in column_indices:
+        col = columns[c - 1]
+        message += f"`{col}`"
```
Comment (neelasha23) [inline src/sql/inspect.py:119]: Please use format strings instead of `+`

File: src/sql/inspect.py (hunk 2 above)
Comment (neelasha23) [reply src/sql/inspect.py:81]: What if the column is categorical and has values like 1,2, ..etc. This function would still return True right?

File: src/sql/inspect.py [hunk 4]
```diff
@@ -77,6 +77,51 @@ def _get_row_with_most_keys(rows):
...
+        pass
+
+
+def _generate_column_styles(
+    column_indices, background_color="#FFFF00", text_color="black"
+):
+    """Change the background-color of all columns with data-type mismatch"""
+    styles = ""
+    for index in column_indices:
+        styles += f"""
```
Comment (neelasha23) [inline src/sql/inspect.py:105]: Use format strings instead of `+`

File: src/sql/inspect.py [hunk 5]
```diff
@@ -149,10 +194,31 @@ def __init__(self, table_name, schema=None) -> None:
...
 
+            # check the datatype of a column
+            try:
+                result = sql.run.raw_run(
+                    Connection.current, f"""SELECT {column} FROM {table_name} LIMIT 1"""
+                ).fetchone()
+
+                value = result[0]
+                if isinstance(value, (int, float)) or (
+                    isinstance(value, str) and _is_numeric(value)
```
Comment (neelasha23) [inline src/sql/inspect.py:211]:
```is_numeric = if isinstance(value, (int, float)) or (
                    isinstance(value, str) and _is_numeric(value))
```
should have the similar effect

File: src/sql/inspect.py [hunk 6]
```diff
@@ -149,10 +194,31 @@ def __init__(self, table_name, schema=None) -> None:
...
+                ):
+                    is_numeric = True
+                else:
+                    is_numeric = False
+            except Exception:
```
Comment (neelasha23) [inline src/sql/inspect.py:216]: Add specific exceptions

File: src/sql/inspect.py [hunk 7]
```diff
@@ -77,6 +77,51 @@ def _get_row_with_most_keys(rows):
...
+            return True
+        return False
+    except ValueError:
+        pass
+
+
+def _generate_column_styles(
+    column_indices, background_color="#FFFF00", text_color="black"
+):
+    """Change the background-color of all columns with data-type mismatch"""
```
Comment (neelasha23) [inline src/sql/inspect.py:102]: Add docstring instead of comment

File: src/sql/inspect.py [hunk 8]
```diff
@@ -162,8 +228,12 @@ def __init__(self, table_name, schema=None) -> None:
                     GROUP BY top ORDER BY frequency Desc""",
                 ).fetchall()
 
-                table_stats[column]["freq"] = result_col_freq_values[0][1]
-                table_stats[column]["top"] = result_col_freq_values[0][0]
+                if is_numeric is False:
+                    table_stats[column]["freq"] = result_col_freq_values[0][1]
+                    table_stats[column]["top"] = result_col_freq_values[0][0]
+                else:
+                    table_stats[column]["freq"] = math.nan
```
Comment (neelasha23) [inline src/sql/inspect.py:235]: If column is numeric type we are setting frequency and top = NaN. Do we know why `pandas` does this?

File: src/sql/inspect.py [hunk 9]
```diff
@@ -220,7 +292,12 @@ def __init__(self, table_name, schema=None) -> None:
                                 """,
                 ).fetchall()
 
-                table_stats[column]["mean"] = float(results_avg[0][0])
+                if is_numeric:
+                    table_stats[column]["mean"] = format(
```
Comment (neelasha23) [inline src/sql/inspect.py:296]: There are multiple code blocks like :

```
if is_numeric:
 ...
else:
  ...
```

Can we consolidate these in a couple of functions - one for `is_numeric = true` other for False

**Comment by neelasha23** [comment]
A few observations:

1. Please document this.
2. Where are we testing the background styles/ colors, etc?
3. How should I interpret the results on the `penguins` dataset? I think it's still not clear to me why these nan values have come up.
Screenshot 2023-06-16 at 11 26 46 AM

4. It looks like currently we are highlighting the column which is numeric but stored as string, and displaying a message. What about highlighting when `stdev` is not supported? : https://github.com/ploomber/jupysql/issues/459#issuecomment-1525751997

File: CHANGELOG.md [hunk 10]
```diff
@@ -2,6 +2,7 @@
 
 ## 0.7.9dev
 
+* [Feature] Modified `TableDescription` class to improve table profiling (#459)
```
Comment (neelasha23) [inline CHANGELOG.md:5]: Please include the exact modifications/formattings added

File: src/tests/test_magic_cmd.py [hunk 11]
```diff
@@ -168,12 +168,69 @@ def test_table_profile(ip, tmp_empty):
...
+    out = ip.run_cell("%sqlcmd profile -t numbers").result
+
+    stats_table = out._table
+
+    assert len(stats_table.rows) == len(expected)
+
+    for row in stats_table:
+        criteria = row.get_string(fields=[" "], border=False).strip()
+
+        rating = row.get_string(fields=["rating"], border=False, header=False).strip()
```
Comment (neelasha23) [inline src/tests/test_magic_cmd.py:188]: There can be a common function `_get_row_string(column_name)` and use this to retrieve each row

File: src/tests/test_magic_cmd.py [hunk 12]
```diff
@@ -168,12 +168,69 @@ def test_table_profile(ip, tmp_empty):
...
+    }
+
+    out = ip.run_cell("%sqlcmd profile -t numbers").result
+
+    stats_table = out._table
+
+    assert len(stats_table.rows) == len(expected)
+
+    for row in stats_table:
+        criteria = row.get_string(fields=[" "], border=False).strip()
```
Comment (neelasha23) [inline src/tests/test_magic_cmd.py:186]: `criteria` -> `profile_metric`

File: src/sql/inspect.py (hunk 2 above)
Comment (AnirudhVIyer) [reply src/sql/inspect.py:81]: If a column only has numerical values then it is not a categorical column. 
if it contains values like 1,2,3 stored as a string then it is still a numeric column so `_is_numeric` will be true,
 but  the` _is_numeric_as_str` will give a message saying that there is a data mismatch

File: src/sql/inspect.py (hunk 3 above)
Comment (AnirudhVIyer) [reply src/sql/inspect.py:119]: making the changes!

File: src/sql/inspect.py (hunk 8 above)
Comment (AnirudhVIyer) [reply src/sql/inspect.py:235]: So frequency and top are usually stats reserved for categorical data. 
If i have a column with 5 unique values `car, bike plane cycle, bus`.
Top indicates the value that has appeared the most in the data 
freq indicated how many times the `top` value appears in the data.

These stats don't translate to numerical data because in a numerical column, two data points usually won't be the same. Most of the data points will be unique, so `Top` and `Freq` are set to Nan.

File: src/sql/inspect.py (hunk 9 above)
Comment (AnirudhVIyer) [reply src/sql/inspect.py:296]: created a function `_assign_column_specific_stats` that takes in ` table_stats[column] ` and `is_numeric` and does what is done in this block : 
```
if is_numeric:
 ...
else:
  ...
```

Now we won't see those multiple blocks in code.

File: src/tests/test_magic_cmd.py (hunk 11 above)
Comment (AnirudhVIyer) [reply src/tests/test_magic_cmd.py:188]: made this change added a helper function `_get_row_string(column_name)` that is used in three different pytest functions.

File: src/sql/inspect.py (hunk 2 above)
Comment (AnirudhVIyer) [reply src/sql/inspect.py:81]: added test `test_table_profile_data_mismatch` to check if a message and styling is applied to a column with data mismatch. `is_numeric` is an internal function being used in every column to determine the type of column it is. I think it is being verified in the existing tests and we don't need a separate test for it.

**Comment by AnirudhVIyer** [comment]
> A few observations:
> 
> 1. Please document this.
> 2. Where are we testing the background styles/ colors, etc?
> 3. How should I interpret the results on the `penguins` dataset? I think it's still not clear to me why these nan values have come up.
> 
> Screenshot 2023-06-16 at 11 26 46 AM
> 4. It looks like currently we are highlighting the column which is numeric but stored as string, and displaying a message. What about highlighting when `stdev` is not supported? : [sqlcmd profile improvements #459 (comment)](https://github.com/ploomber/jupysql/issues/459#issuecomment-1525751997)


Added a message when using any database other than `duckdb` that STD, 25h, 50h and 75h percentiles won't be available. 

Added a test that checks styling and warnings. `test_table_profile_warnings_styles`

Also updated docstring making it clear.

**Comment by AnirudhVIyer** [comment]
Made the required changes 
 
Screenshot 2023-06-16 at 5 17 23 PM

Now we add styling to the mismatched datatype column, issue a warning. 

If not using duckdb, issue a warning about the available statistics. 

Added required tests as well.

**Comment by tonykploomber** [comment]
Can we have a preview of how it looks like in dark theme in jupyter? Since I am using that theme now

**Comment by AnirudhVIyer** [comment]
> Can we have a snapshot preview of how it looks like in dark theme in jupyter? Since I am using that theme now (AC: just make sure it looks not too bad :) )

just tried it, looks decent imo 😂
Screenshot 2023-06-16 at 6 59 05 PM

File: src/sql/inspect.py (hunk 2 above)
Comment (neelasha23) [reply src/sql/inspect.py:81]: Ok. This part is not very clear to me. 
As per my understanding, categorical columns can be either numeric or strings. So in one dataset `sex` can be represented as `Male` and `Female`, in some other dataset it can be represented as `1` and `2`.
Are we assuming the latter case will never happen?

We add testcase for internal functions as well if the function is significant. Especially in this case it'll help to understand the function better through different testcases.

File: src/sql/inspect.py [hunk 13]
```diff
@@ -268,39 +376,94 @@ def __init__(self, table_name, schema=None) -> None:
...
+            warning_background = "#FFFFCC"
+            warning_title = "Warning: "
+        else:
+            message_content = ""
+            warning_background = "white"
+            warning_title = ""
 
-            self._table.add_row(values)
+        database = Connection.current.url
+        if "duckdb" in database:
```
Comment (neelasha23) [inline src/sql/inspect.py:425]: Can we utilize this method here? Also, we don't need to mention `only available with dubckdb`. Just mention that these are not available in `{db_driver}`
https://github.com/ploomber/jupysql/blob/a5790a62759310d37abb9d0b6354bae3cdc74a0a/src/sql/util.py#L246

File: src/tests/test_magic_cmd.py [hunk 14]
```diff
@@ -14,6 +14,20 @@
 )
 
 
+def _get_row_string(row, column_name):
+    """
+    Helper function to retrieve the string value of a specific column in a table row.
+
+    Args:
```
Comment (neelasha23) [inline src/tests/test_magic_cmd.py:21]: `Args` -> `Parameters`. You can look at some existing functions to check the convention

File: src/sql/inspect.py (hunk 13 above)
Comment (neelasha23) [reply src/sql/inspect.py:425]: Can we add some integration tests here as well?

Comment (AnirudhVIyer) [reply src/sql/inspect.py:425]: Sure will add them.

File: src/sql/inspect.py (hunk 2 above)
Comment (AnirudhVIyer) [reply src/sql/inspect.py:81]: Sure, I'll add the testcase for the internal function. 

In the categorical column part, we have tried to replicate the pandas `description` function. So any time the categorical classes are represented in their encoded forms `eg. male and female represented as 0 1`, we treat them as numerical.  I think pandas does it because if a category has many classes which are in the encoded form, it will be tough to identify it as categorical. So we end up treating them as numerical columns.

**Comment by neelasha23** [comment]
There are conflicts in the PR. Please rebase

**Comment by edublancas** [comment]
@AnirudhVIyer: I see you fixed the merging conflicts. Is this ready for review again?

**Comment by AnirudhVIyer** [comment]
> @AnirudhVIyer: I see you fixed the merging conflicts. Is this ready for review again?

@edublancas No, I need to add a few integration tests.

**Comment by AnirudhVIyer** [comment]
Added a test for internal function `is_numeric`, added integration tests.

File: src/sql/inspect.py (hunk 13 above)
Comment (AnirudhVIyer) [reply src/sql/inspect.py:428]: Made the changes

File: src/tests/test_magic_cmd.py [hunk 15]
```diff
@@ -227,27 +291,62 @@ def test_table_schema_profile(ip, tmp_empty):
...
+    """
+    )
+    out = ip.run_cell("%sqlcmd profile -t numbers").result
+    stats_table_html = out._table_html
+    assert "Columns `price` have a datatype mismatch" in stats_table_html
+    assert "#profile-table td:nth-child(3" in stats_table_html
+    assert "Following statistics are not available in" in stats_table_html
+
+
+def test_table_profile_is_numeric(ip, tmp_empty):
```
Comment (neelasha23) [inline src/tests/test_magic_cmd.py:333]: This is still testing the overall functionality and not calling `_is_numeric` directly. What I meant was you can add different test cases like integer, integer as string, non integer (string, null value) etc, basically all possible values that can occur in the dataset and test `_is_numeric` directly

Comment (AnirudhVIyer) [reply src/tests/test_magic_cmd.py:333]: Ok, got it!

Comment (AnirudhVIyer) [reply src/tests/test_magic_cmd.py:345]: Added the test function `test_profile_is_numeric` that implements the above changes.

**Comment by AnirudhVIyer** [comment]
@edublancas ready for review

**Review by edublancas — CHANGES_REQUESTED** [review]
I tried testing this and I was expecting the island column to be yellow and to see the warning message since island is a string: 

image

(using the penguins.csv dataset we download in the quick start)

am I doing something wrong?

please share the data you used for your tests so I can check it out

**Comment by AnirudhVIyer** [comment]
> I tried testing this and I was expecting the island column to be yellow and to see the warning message since island is a string:
> 
> image
> (using the penguins.csv dataset we download in the quick start)
> 
> am I doing something wrong?
> 
> please share the data you used for your tests so I can check it out

Hey, so the island column in the penguins dataset is made up of strings and is a categorical column(is not bade up of integers), we show the column to be yellow and add a warning message if the column is made of strings but contains numerical values (like ints/floats). Because of this we cannot properly query the statistics through SQL.

In case of the churn.csv file the `totalCharges` column represents integer values but they are stored as strings. 
image

you can test it on either churn.csv file or this mock dataset:

`
%%sql sqlite://
CREATE TABLE people (name varchar(50),age varchar(50),number int,
    country varchar(50),gender_1 varchar(50), gender_2 varchar(50));
INSERT INTO people VALUES ('joe', '48', 82, 'usa', '0', 'male');
INSERT INTO people VALUES ('paula', '50', 93, 'uk', '1', 'female');
`

`%sqlcmd profile -t people`

**Review by edublancas — CHANGES_REQUESTED** [review]
ok, I think I did something wrong because now when I profile penguins.csv I see it in yellow:

image

suggestion: can we make the yellow background softer? (the current one is too strong) the yellow one you show in the warning looks better, let's highlight the columns with the same color:

image


also, note that the warning shows. 

~~~
`gender` <- backticks
~~~

can we change the backticks for something else? perhaps we can use

```
<code>column_name</code>
```

**Comment by AnirudhVIyer** [comment]
> ok, I think I did something wrong because now when I profile penguins.csv I see it in yellow:
> 
> image
> 
> suggestion: can we make the yellow background softer? (the current one is too strong) the yellow one you show in the warning looks better, let's highlight the columns with the same color:
> 
> image
> 
> also, note that the warning shows.
> 
> ```
> `gender` <- backticks
> ```
> 
> can we change the backticks for something else? perhaps we can use
> 
> ```
> <code>column_name</code>
> ```

Making the changes to the styling, don't know why penguins.csv has the yellow columns now, runs fine locally. I will check it out.

**Comment by edublancas** [comment]
> Making the changes to the styling, don't know why penguins.csv has the yellow columns now, runs fine locally. I will check it out.

I think it's **good** that the island column in the penguins dataset is displayed on yellow, since we can't compute the statistics on strings. so let's keep it (if it's a bug, let's make it a feature 😂)

having said that, we can also display a message like:

```
the island column is a string type. cannot calculate a,b,c,d...
```

**Review by edublancas — APPROVED** [review]
